packages/next/src/client/components/router-reducer/reducers/prefetch-reducer.ts

```diff
@@ -12,15 +11,29 @@ import {
 
 export const prefetchQueue = new PromiseQueue(5)
 
-export function prefetchReducer(
+export const prefetchReducer =
+  process.env.__NEXT_PPR && process.env.__NEXT_CLIENT_SEGMENT_CACHE
+    ? identityReducerWhenSegmentCacheIsEnabled
+    : prefetchReducerImpl
+
+function identityReducerWhenSegmentCacheIsEnabled<T>(state: T): T {
+  // Unlike the old implementation, the Segment Cache doesn't store its data in
+  // the router reducer state.
+  //
+  // This shouldn't be reachable because we wrap the prefetch API in a check,
+  // too, which prevents the action from being dispatched. But it's here for
+  // clarity + code elimination.
+  return state
+}
+
+function prefetchReducerImpl(
   state: ReadonlyReducerState,
   action: PrefetchAction
 ): ReducerState {
   // let's prune the prefetch cache before we do anything else
   prunePrefetchCache(state.prefetchCache)
 
   const { url } = action
-  url.searchParams.delete(NEXT_RSC_UNION_QUERY)
```

Comment:
Agree that it doesn't appear to be needed - the query param is added at the `fetchServerResponse` phase which comes after the prefetch action is dispatched anyway. I suspect it was just a greedy find and replace for all things related to `url.searchParams`.

Reply:
Note: This accidentally fixed https://github.com/vercel/next.js/issues/72927

Because the invocation of methods on `url.searchParams` actually mutates the URL and escapes percent-encoding.

```
u = new URL("http://localhost:3000/?q=a%20b");

u.toString(); // 'http://localhost:3000/?q=a%20b'

u.searchParams.append("p", 1); // any u.searchParams method invocation

u.toString(); // 'http://localhost:3000/?q=a+b&p=1'

// `%20` turned to `+`
```